sqlx-postgres/src/types/jiff/datetime.rs

```diff
@@ -0,0 +1,127 @@
+use crate::error::BoxDynError;
+use crate::types::Type;
+use crate::{PgArgumentBuffer, PgHasArrayType, PgTypeInfo, PgValueFormat, PgValueRef, Postgres};
+use jiff::tz::TimeZone;
+use jiff::{SignedDuration, Timestamp, Zoned};
+use std::mem;
+use std::str::FromStr;
+
+impl Type<Postgres> for Timestamp {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::TIMESTAMP
+    }
+}
+
+impl Type<Postgres> for Zoned {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::TIMESTAMPTZ
+    }
+}
+
+impl PgHasArrayType for Timestamp {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::TIMESTAMP_ARRAY
+    }
+}
+
+impl PgHasArrayType for Zoned {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::TIMESTAMPTZ_ARRAY
+    }
+}
+
+impl Encode<'_, Postgres> for Timestamp {
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> Result<IsNull, BoxDynError> {
+        // TIMESTAMP is encoded as the microseconds since the epoch
+        let micros = (*self - postgres_epoch_datetime()).get_microseconds();
+        Encode::<Postgres>::encode(micros, buf)
+    }
+
+    fn size_hint(&self) -> usize {
+        mem::size_of::<i64>()
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for Timestamp {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
+        Ok(match value.format() {
+            PgValueFormat::Binary => {
+                // TIMESTAMP is encoded as the microseconds since the epoch
+                let us = Decode::<Postgres>::decode(value)?;
+                postgres_epoch_datetime() + SignedDuration::from_micros(us)
+            }
+            PgValueFormat::Text => {
+                let s = value.as_str()?;
+                parse_timestamp_text(s)?.timestamp()
+            }
+        })
+    }
+}
+
+impl Encode<'_, Postgres> for Zoned {
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> Result<IsNull, BoxDynError> {
+        Encode::<Postgres>::encode(self.timestamp(), buf)
+    }
+
+    fn size_hint(&self) -> usize {
+        mem::size_of::<i64>()
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for Zoned {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
+        Ok(match value.format() {
+            PgValueFormat::Binary => {
+                let naive = <Timestamp as Decode<Postgres>>::decode(value)?;
+                naive.to_zoned(TimeZone::UTC)
+            }
+            PgValueFormat::Text => {
+                let s = value.as_str()?;
+                parse_timestamp_text(s)?
+            }
+        })
+    }
+}
+
+#[inline]
+fn parse_timestamp_text(input: &str) -> Result<Zoned, BoxDynError> {
+    Ok(Zoned::strptime(
+        if input.contains('+') || input.contains('-') {
+            // Contains a time-zone specifier
+            // This is given for timestamptz for some reason
+            // Postgres already guarantees this to always be UTC
+            "%Y-%m-%d %H:%M:%S%.f%#z"
+        } else {
+            "%Y-%m-%d %H:%M:%S%.f"
+        },
+        input,
+    )?)
+}
```

Comment:
This is incorrect. But I don't find a workaround now.

`chrono` can parse "2004-10-19 10:23:54+00" but jiff will failed with:

```
Err(Error { inner: ErrorInner { kind: Adhoc(strptime parsing failed), cause: Some(Error { inner: ErrorInner { kind: Adhoc(%z failed), cause: Some(Error { inner: ErrorInner { kind: Adhoc(expected at least 4 digits for time zone offset after sign, but found only 2 bytes remaining), cause: None } }) } }) } })
```

BTW, `if input.contains('+') || input.contains('-')` doesn't seems useful because `%Y-%m-%d` always contains `-`.

Reply:
In short,

* chrono's DateTime supports parsing `%.f%#z` without precision and timezone in short form
* chrono's NaiveDateTime supports parsing without timezone, but jiff's Timestamp seems cannot (?)

Reply:
Maybe the root is what text will Postgres return and we can think of how to process it.

Reply:
Pushed https://github.com/launchbadge/sqlx/pull/3511/commits/3e5f5d06bf407e3d6000f6f24f479939ca945f62 to rework it.

The final issue is that jiff's `strptime` `%#z` support only `+0200` but not `+02` which seems can be returned by postgres.

Reply:
You're running into issues here because you're kinda doing something you're not supposed to be doing. A `Zoned` is a timestamp _with a time zone_. But you don't have a time zone in this context. All you have is a timestamp and an offset.

If you just use `Timestamp`'s `FromStr` impl, then it will parse this format just fine. Including `+02`, `+02:00`, `+0200` and so on.

You'll notice that adding Jiff support to `rust-postgres` completely omitted `Zoned` for this reason: https://github.com/sfackler/rust-postgres/pull/1164

Reply:
Thank you! Seems I should use Timestamp instead here.

Reply:
Updated.